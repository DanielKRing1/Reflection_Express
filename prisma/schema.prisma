// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_LOCAL")
}

model User {
  password_userId             Password?           @relation(name: "password_userId")
  journal_userId              Journal[]           @relation(name: "journal_userId")
  email                       String              @id @unique

  // Set to null when delete Journal + Update id if Journal id updated
  user_lastUsedJournalId      Journal?            @relation(name: "user_lastUsedJournalId", fields: [lastUsedJId], references: [id], onDelete: SetNull, onUpdate: Cascade)
  lastUsedJId                 BigInt?             @unique
}

model Password {
  // Cascade delete/update with User
  password_userId             User                @relation(name: "password_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String              @id

  hash                        String
}

model Journal {
  user_lastUsedJournalId      User?               @relation(name: "user_lastUsedJournalId")
  inkling_journalId           Inkling[]           @relation(name: "inkling_journalId")
  journalentry_journalId      JournalEntry[]      @relation(name: "journalentry_journalId")
  thought_journalId           Thought[]           @relation(name: "thought_journalId")
  reflection_journalId        Reflection[]        @relation(name: "reflection_journalId")
  id                          BigInt              @id @default(autoincrement())
  
  // Cascade delete/update with User
  journal_userId              User                @relation(name: "journal_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String

  @@index([userId])
}

model Inkling {
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  inkling_journalId           Journal             @relation(name: "inkling_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model JournalEntry {
  reflection_journalEntryId   Reflection[]        @relation(name: "reflection_journalEntryId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  journalentry_journalId      Journal             @relation(name: "journalentry_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  @@id([journalId, timeId])
}

model Thought {
  reflection_thoughtId        Reflection[]        @relation(name: "reflection_thoughtId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  thought_journalId           Journal             @relation(name: "thought_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model Reflection {
  // Cascade delete/update with Journal
  reflection_journalId        Journal             @relation(name: "reflection_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  // Cascade delete/update with JournalEntry
  reflection_journalEntryId   JournalEntry        @relation(name: "reflection_journalEntryId", fields: [journalId, journalEntryId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  journalEntryId              DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Thought
  reflection_thoughtId        Thought             @relation(name: "reflection_thoughtId", fields: [journalId, thoughtId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  thoughtId                   DateTime            @db.Timestamptz(3)

  decision                    Int                 @db.SmallInt

  @@id([journalEntryId, thoughtId])
}
