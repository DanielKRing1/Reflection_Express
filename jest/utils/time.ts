export const useFakeTimer = (startDate: Date) =>
    jest
        .useFakeTimers({
            doNotFake: [
                "nextTick",
                "setImmediate",
                "clearImmediate",
                "setInterval",
                "clearInterval",
                "setTimeout",
                "clearTimeout",
            ],
        })
        .setSystemTime(startDate);

export const advanceFakeTimer = (msToRun: number) =>
    jest.advanceTimersByTime(msToRun);

export const useRealTimer = () => jest.useRealTimers();
