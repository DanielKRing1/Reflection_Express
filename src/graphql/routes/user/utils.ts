import { Request } from "express";
import { GQL_PATH } from "../../constants";
import { AUTH_OP_NAMES } from "./schema.gql";

export const isGqlAuthRequest = (req: Request): boolean => {
  // 1. Is not graphql route
  const fullPath = req.baseUrl + req.path;
  if (fullPath !== GQL_PATH) return false;

  // 2. Is not auth operation
  const opName: string = req.body.operationName.toLocaleLowerCase();
  if (!AUTH_OP_NAMES.has(opName)) return false;

  return true;
};
