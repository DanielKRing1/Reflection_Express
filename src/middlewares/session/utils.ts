import session from "express-session";
import { createClient } from "redis";
import RedisStore from "connect-redis";

type CreateRedisSession = {
  redisPrefix: string;
} & Omit<CreateSession<RedisStore>, "sessionStore">;
/**
 * Create express-session with Redis store
 *
 * @param param0
 * @returns
 */
export const createRedisSession = ({
  redisPrefix,
  sessionName,
  secret,
  maxAge,
  cookiePath,
}: CreateRedisSession) => {
  const redisStore: RedisStore = createRedisStore(redisPrefix);
  return createSession({
    sessionName,
    sessionStore: redisStore,
    secret,
    maxAge,
    cookiePath,
  });
};

/**
 * Connect to Redis store
 *
 * @param prefix
 * @returns
 */
const createRedisStore = (prefix: string): RedisStore => {
  let redisClient = createClient();
  redisClient.connect().catch(console.error);

  return new RedisStore({
    client: redisClient,
    prefix,
  });
};

type CreateSession<T> = {
  sessionName: string;
  sessionStore?: T;
  secret: string;
  maxAge: number;
  cookiePath: string;
};
/**
 * Create express-session
 *
 * @param param0
 * @returns
 */
export const createSession = ({
  sessionName,
  sessionStore,
  secret,
  maxAge,
  cookiePath,
}: CreateSession<any>) => {
  return session({
    secret,
    store: sessionStore,
    name: sessionName,
    resave: false, // Do not re-submit 'set' action to store when 'session' obj has not been modified
    saveUninitialized: false, // Do not 'create' session in store when 'session' obj has not been modified
    cookie: {
      path: cookiePath,
      // secure: true, // if true only transmit cookie over https
      httpOnly: true, // if true prevent client side JS from reading the cookie
      maxAge,
    },
  });
};
