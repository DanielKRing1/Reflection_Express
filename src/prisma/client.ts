import { PrismaClient } from "./generated/client";

let prisma = new PrismaClient({
    // log: ["query"],
});

export default prisma;
