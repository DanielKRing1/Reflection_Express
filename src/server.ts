// EXPRESS
import express, { Express } from "express";
import cors from "cors";

// MIDDLEWARE
import loggingMiddleware from "./middlewares/logging.middleware";

// ROUTES
import genRouter from "./routes";

// CONFIG
import { PORT } from "./config/server.config";

// APP
const app: Express = express();

// MIDDLEWARE
app.use(cors());
// parse requests of content-type - application/json
app.use(express.json());
// parse requests of content-type - application/x-www-form-urlencoded
app.use(
  express.urlencoded({
    extended: true,
  })
);
app.use(loggingMiddleware);

// GRAPHQL
app.use(genRouter());

// START SERVER
app.listen(PORT, () => console.log(`Server listening on port ${PORT}!`));
