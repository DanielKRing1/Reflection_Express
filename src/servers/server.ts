import genExpressApp from "./app";
import gqlServer from "./gqlServer";

import accessPromise from "../middlewares/session/access";

export default (async () => {
    const app = await genExpressApp();
    const access = await accessPromise();

    return gqlServer(app, access);
})();
