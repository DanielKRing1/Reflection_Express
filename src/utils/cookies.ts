import { Response } from "express";
import setCookie from "set-cookie-parser";

export type HttpCookieResponse = {
  status: number;
  headers: {
    "set-cookie": string;
  } & any;
};
export const mergeCookies = (
  httpResponse: HttpCookieResponse,
  expressRes: Response
) => {
  if (httpResponse.status >= 400)
    throw new Error(
      `mergeCookies() received an httpResponse with status code: ${httpResponse.status}`
    );

  // 6.3. Strip refresh cookie from response
  // Get all cookies from cookie header
  const cookies = setCookie.parse(httpResponse as any);
  console.log(cookies);

  // 6.4. Attach refresh cookie to res.cookie
  cookies.forEach((cookie) =>
    expressRes.cookie(cookie.name, cookie.value, cookie as any)
  );
};
