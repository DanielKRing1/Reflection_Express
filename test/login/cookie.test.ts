import http from "http";
import supertest from "supertest";

import setCookie from "set-cookie-parser";

import server from "../../src/index";
import prisma from "../../src/prisma/client";
import {
    ACCESS_SESSION_COOKIE_NAME,
    META_ACCESS_SESSION_COOKIE_NAME,
} from "../../src/middlewares/session/access/constants";
import {
    REFRESH_SESSION_COOKIE_NAME,
    META_REFRESH_SESSION_COOKIE_NAME,
} from "../../src/middlewares/session/refresh/constants";
import { COOKIE_DOMAIN } from "../../src/utils/cookies";
import { useFakeTimer } from "../../jest/utils/time";

const USER_ID = "1";
const PASSWORD = "123";

const PRESENT_DATE = new Date(2023, 3, 16);

describe("User GraphQL server", () => {
    let s: http.Server;
    let agent: supertest.SuperAgentTest;

    beforeAll(async () => {
        console.log("start beforeAll");
        // Mock Date constructor
        console.log("going beforeAll");

        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;
        console.log("after query beforeAll");

        s = await server;
        console.log("after start server beforeAll");
        agent = supertest.agent(s);

        useFakeTimer(PRESENT_DATE);
    });

    it("Should get access and refresh session cookies (at least 2 cookies)", async () => {
        console.log("start test1");
        const res = await agent
            .post("/login/create-user")
            .send({ userId: USER_ID, password: PASSWORD });
        console.log("going test1");

        const cookies = setCookie
            .parse(res as any)
            .sort((a, b) => a.name.localeCompare(b.name));
        console.log(cookies);
        expect(cookies.length).toBe(4);

        expect(cookies[0].name).toBe(ACCESS_SESSION_COOKIE_NAME);
        expect(cookies[1].name).toBe(META_ACCESS_SESSION_COOKIE_NAME);
        expect(cookies[2].name).toBe(REFRESH_SESSION_COOKIE_NAME);
        expect(cookies[3].name).toBe(META_REFRESH_SESSION_COOKIE_NAME);

        console.log(
            agent.jar.getCookies({
                domain: COOKIE_DOMAIN,
                path: "/refresh",
                secure: false,
                script: false,
            })
        );
    });

    it("Should get the created User (this requires successfully sending an access session cookie with the request)", async () => {
        const queryData = {
            query: `query User {
                user {
                    email
                    lastUsedJId
                }
            }`,
            variables: { email: USER_ID },
        };

        console.log("start test2");
        const res = await agent.post("/graphql").send(queryData);
        console.log("going test2");

        expect(res.text).toMatchSnapshot();
    });

    afterAll(async () => {
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;

        await prisma.$disconnect();
        await s.close();
    });
});
