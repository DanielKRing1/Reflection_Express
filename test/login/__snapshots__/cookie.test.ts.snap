// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User GraphQL server Should get the created User (this requires successfully sending an access session cookie with the request) 1`] = `
"{"data":{"user":{"email":"1","lastUsedJId":null}}}
"
`;
